contracts/FIND.cdc

```diff
@@ -176,30 +176,30 @@ pub contract FIND {
 		
 		let profileFindName= Profile.find(address).getFindName()
 
+		let network = self.account.borrow<&Network>(from: FIND.NetworkStoragePath) ?? panic("Network is not set up")
+
 		if profileFindName != "" {
-			if let network = self.account.borrow<&Network>(from: FIND.NetworkStoragePath) {
-				if network.readStatus(profileFindName).status == FIND.LeaseStatus.TAKEN {
-					return profileFindName
-				}
+			if network.readStatus(profileFindName).status == FIND.LeaseStatus.TAKEN {
+				return profileFindName
 			}
 		}
 
-		let leases = leaseCap.borrow()!.getLeaseInformation() 
+		let leaseCol = leaseCap.borrow()!
+		let nameLeases = leaseCol.getNames() 
 		var time : UFix64?= nil
 		var name :String?= nil
-		for lease in leases {
+		for nameLease in nameLeases {
 
 			//filter out all leases that are FREE or LOCKED since they are not actice
-			if lease.status != "TAKEN" {
+			if network.readStatus(nameLease).status != FIND.LeaseStatus.TAKEN {
```

Comment:
I think we need to change the algorithm here. Accept the first name that is valid and not the first one registered.

Reply:
It is too expensive to continue doing this.